### PR TITLE
Azure AD entities description

### LINKED ISSUE #875: App Roles for the App Registration should be documented
I had to dig through the python deployment script in order to understand the app roles that needed to be configured on the app registration.

https://github.com/microsoft/onefuzz/blob/54d49a973e4007b3db2ff0d134d368c648183524/src/deployment/deploy.py#L272-L289

## Provide a description of requested docs changes

This should more clearly be delineated in markdown documentation in the event that someone is trying to create the app registration by hand. In most secure enterprises, people do not have standing access to create apps / service principals in Azure AD and there is typically a process for getting one created. In this event, in would be helpful to have this documented to save people time from trying to discern all the things the python deployment script is doing.

@shambho 
The deployment script is the only supported way to register the application. We do not support the manual creation. 
You are welcome to translate the deployment code to manual steps or wait for the audit of the deployment script.

_Originally posted by @chkeita in https://github.com/microsoft/onefuzz/issues/813#issuecomment-827229205_

A few lines of documentation on how to set up the app registration is a reasonable ask here.

### PR DESCRIPTION
**Pull request by chkeita**
Added a description of the configuration of azure AD entities

closes #875

### CODE REVIEW

File: docs/AADEntitites.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# Azure Ad Entities
+This document describes the configuration of entities create in azure AD by our [deployment script](../src/deployment/deploy.sh)
```
Comment (ranweiler) [inline docs/AADEntitites.md:2]:
```suggestion
This document describes the configuration of entities create in Azure AD by our [deployment script](../src/deployment/deploy.sh)
```

File: docs/AADEntitites.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# Azure Ad Entities
```
Comment (ranweiler) [inline docs/AADEntitites.md:1]:
```suggestion
# Azure Active Directory Entities
```

File: docs/AADEntitites.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+This document describes the configuration of entities create in azure AD by our [deployment script](../src/deployment/deploy.sh)
+
+### Onefuzz Application registration
```
Comment (ranweiler) [inline docs/AADEntitites.md:4]:
```suggestion
### OneFuzz Application Registration
```

File: docs/AADEntitites.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+This is the registration of the onefuzz instance
```
Comment (ranweiler) [inline docs/AADEntitites.md:5]:
```suggestion
This is the registration of the OneFuzz instance:
```

File: docs/AADEntitites.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        * value: ManagedNode
+        * Allowed Member types: Applications
+    * CliClient
+        * value: ManagedNode
+        * Allowed Member types: Applications
+* Api Permissions
+    * User.Read (Microsoft Graph)
+* scope
+    * user_impersonation
+* Authorozed application:
```
Comment (ranweiler) [inline docs/AADEntitites.md:18]:
```suggestion
* Authorized application:
```

File: docs/AADEntitites.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    * Onefuzz cli registration
```
Comment (ranweiler) [inline docs/AADEntitites.md:19]:
```suggestion
    * OneFuzz CLI registration
```

File: docs/AADEntitites.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+### Onefuzz Application Service Principal
+service principal linked to the onefuzz application registration
```
Comment (ranweiler) [inline docs/AADEntitites.md:22]:
```suggestion
Service principal linked to the OneFuzz application registration:
```

File: docs/AADEntitites.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+* name: <instance_name>
+* Application Id: <Onefuzz Application registration app_id>
+
+### Onefuzz cli registration
```
Comment (ranweiler) [inline docs/AADEntitites.md:26]:
```suggestion
### OneFuzz CLI registration
```

File: docs/AADEntitites.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    * User.Read (Microsoft Graph)
+* scope
+    * user_impersonation
+* Authorozed application:
+    * Onefuzz cli registration
+
+### Onefuzz Application Service Principal
+service principal linked to the onefuzz application registration
+* name: <instance_name>
+* Application Id: <Onefuzz Application registration app_id>
```
Comment (ranweiler) [inline docs/AADEntitites.md:24]:
```suggestion
* Application Id: <OneFuzz Application registration app_id>
```

File: docs/AADEntitites.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+### Onefuzz cli registration
+The registration for the Command line interface
```
Comment (ranweiler) [inline docs/AADEntitites.md:27]:
```suggestion
The registration for the command line interface:
```

File: docs/AADEntitites.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+* name: <instance_name>-cli
+
+### Onefuzz cli Service Principal
```
Comment (ranweiler) [inline docs/AADEntitites.md:30]:
```suggestion
### OneFuzz CLI Service Principal
```

File: docs/AADEntitites.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+service principal linked to the onefuzz cli application registration
```
Comment (ranweiler) [inline docs/AADEntitites.md:31]:
```suggestion
Service principal linked to the OneFuzz CLI application registration:
```

File: docs/AADEntitites.md [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+* name: <instance_name>-cli
+* Application Id: <Onefuzz cli registration app_id>
```
Comment (ranweiler) [inline docs/AADEntitites.md:33]:
```suggestion
* Application Id: <OneFuzz CLI registration app_id>
```

File: docs/AADEntitites.md [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+* User Assignment required: true
+* Permission
+    * CliClient (from Onefuzz Application registration)
```
Comment (ranweiler) [inline docs/AADEntitites.md:36]:
```suggestion
    * CliClient (from OneFuzz Application registration)
```

File: docs/AADEntitites.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+### Managed node Service Principal
+This entity is available after the first deployment. This is the service principal associated with the user managed identity <instance_name>-scalesetid
```
Comment (ranweiler) [inline docs/AADEntitites.md:39]:
```suggestion
This entity is available after the first deployment. This is the service principal associated with the user-assigned managed identity `<instance_name>-scalesetid`.
```

File: docs/AADEntitites.md [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+* name: <instance_name>-scalesetid
+* Service Principal
+    * Permission
+        * ManagedNode (from Onefuzz Application registration)
```
Comment (ranweiler) [inline docs/AADEntitites.md:44]:
```suggestion
        * ManagedNode (from OneFuzz Application registration)
```

File: docs/AADEntitites.md (hunk 1 above)
Comment (ranweiler) [inline docs/AADEntitites.md:2]:
```suggestion
This document describes the configuration of entities create in azure AD by our [deployment script](../src/deployment/deploy.py).
```

File: docs/AADEntitites.md [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+This is the registration of the onefuzz instance
+* name : <instance_name>
+* app roles
+    * ManagedNode
+        * value: ManagedNode
+        * Allowed Member types: Applications
+    * CliClient
+        * value: ManagedNode
+        * Allowed Member types: Applications
+* Api Permissions
```
Comment (ranweiler) [inline docs/AADEntitites.md:14]:
```suggestion
* API Permissions
```

File: docs/AADEntitites.md [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+* Permission
+    * CliClient (from OneFuzz Application registration)
+
+### Managed node Service Principal
+This entity is available after the first deployment. This is the service principal associated with the user-assigned managed identity `<instance_name>-scalesetid`.
+
+* name: <instance_name>-scalesetid
+* Service Principal
+    * Permission
+        * ManagedNode (from Onefuzz Application registration)
```
Comment (chkeita) [inline docs/AADEntitites.md:44]:
``` Suggestion
        * ManagedNode (from OneFuzz Application registration)
```

File: docs/AADEntitites.md [hunk 19]
```diff
@@ -0,0 +1,44 @@
+# Azure Active Directory Entities
+This document describes the configuration of entities create in Azure AD by our [deployment script](../src/deployment/deploy.sh)
```
Comment (chkeita) [inline docs/AADEntitites.md:2]:
```suggestion
This document describes the configuration of entities create in Azure AD by our [deployment script](../src/deployment/deploy.py)
```

File: docs/AADEntitites.md [hunk 20]
```diff
@@ -0,0 +1,44 @@
+# Azure Active Directory Entities
+This document describes the configuration of entities create in Azure AD by our [deployment script](../src/deployment/deploy.py)
+
+### OneFuzz Application Registration
+This is the registration of the OneFuzz instance:
```
Comment (ranweiler) [inline docs/AADEntitites.md:5]:
```suggestion
This is the registration of the OneFuzz instance.
```

File: docs/AADEntitites.md [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+* name : <instance_name>
+* app roles
+    * ManagedNode
+        * value: ManagedNode
+        * Allowed Member types: Applications
+    * CliClient
+        * value: ManagedNode
+        * Allowed Member types: Applications
+* API Permissions
+    * User.Read (Microsoft Graph)
```
Comment (ranweiler) [inline docs/AADEntitites.md:15]: Suggestion: link to docs, if available.

File: docs/AADEntitites.md (hunk 20 above)
Comment (ranweiler) [reply docs/AADEntitites.md:5]: (I suggested `:` previously, but it actually looks weird in context, sorry about that!)

File: docs/AADEntitites.md [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+* scope
+    * user_impersonation
+* Authorized application:
+    * OneFuzz CLI registration
+
+### Onefuzz Application Service Principal
+Service principal linked to the OneFuzz application registration:
```
Comment (ranweiler) [inline docs/AADEntitites.md:22]:
```suggestion
Service principal linked to the OneFuzz application registration.
```

File: docs/AADEntitites.md [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+* name: <instance_name>
+* Application Id: <OneFuzz Application registration app_id>
+
+### OneFuzz CLI registration
+The registration for the command line interface:
```
Comment (ranweiler) [inline docs/AADEntitites.md:27]:
```suggestion
The registration for the command line interface.
```

File: docs/AADEntitites.md [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+* name: <instance_name>-cli
+
+### OneFuzz CLI Service Principal
+Service principal linked to the OneFuzz CLI application registration:
```
Comment (ranweiler) [inline docs/AADEntitites.md:31]:
```suggestion
Service principal linked to the OneFuzz CLI application registration.
```

File: docs/AADEntitites.md (hunk 21 above)
Comment (ranweiler) [reply docs/AADEntitites.md:15]: (apply this comment to any other standard API permissions that are referenced)

File: docs/AADEntitites.md [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+* name: <instance_name>-cli
+* Application Id: <OneFuzz CLI registration app_id>
+* User Assignment required: true
+* Permission
+    * CliClient (from OneFuzz Application registration)
+
+### Managed node Service Principal
```
Comment (ranweiler) [inline docs/AADEntitites.md:38]:
```suggestion
### Managed Node Service Principal
```

File: docs/AADEntitites.md [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+This entity is available after the first deployment. This is the service principal associated with the user-assigned managed identity `<instance_name>-scalesetid`.
+
+* name: <instance_name>-scalesetid
```
Comment (ranweiler) [inline docs/AADEntitites.md:41]:
```suggestion
* name: `<instance_name>-<scaleset_id>`
```

File: docs/AADEntitites.md [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+* name: <instance_name>
+* Application Id: <OneFuzz Application registration app_id>
+
+### OneFuzz CLI registration
+The registration for the command line interface:
+* name: <instance_name>-cli
+
+### OneFuzz CLI Service Principal
+Service principal linked to the OneFuzz CLI application registration:
+* name: <instance_name>-cli
```
Comment (ranweiler) [inline docs/AADEntitites.md:32]:
```suggestion
* name: `<instance_name>-cli`
```

File: docs/AADEntitites.md [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+    * OneFuzz CLI registration
+
+### Onefuzz Application Service Principal
+Service principal linked to the OneFuzz application registration:
+* name: <instance_name>
+* Application Id: <OneFuzz Application registration app_id>
+
+### OneFuzz CLI registration
+The registration for the command line interface:
+* name: <instance_name>-cli
```
Comment (ranweiler) [inline docs/AADEntitites.md:28]:
```suggestion
* name: `<instance_name>-cli`
```

File: docs/AADEntitites.md [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+* API Permissions
+    * User.Read (Microsoft Graph)
+* scope
+    * user_impersonation
+* Authorized application:
+    * OneFuzz CLI registration
+
+### Onefuzz Application Service Principal
+Service principal linked to the OneFuzz application registration:
+* name: <instance_name>
```
Comment (ranweiler) [inline docs/AADEntitites.md:23]:
```suggestion
* name: `<instance_name>`
```

**Review by ranweiler — APPROVED** [review]
Looks great! Added a few last style suggestions after seeing the result of the last round.

File: docs/AADEntitites.md [hunk 30]
```diff
@@ -0,0 +1,44 @@
+# Azure Active Directory Entities
+This document describes the configuration of entities create in Azure AD by our [deployment script](../src/deployment/deploy.py)
+
+### OneFuzz Application Registration
+This is the registration of the OneFuzz instance.
+* name : <instance_name>
```
Comment (ranweiler) [inline docs/AADEntitites.md:6]:
```suggestion
* name : `<instance_name>`
```

File: docs/AADEntitites.md [hunk 31]
```diff
@@ -0,0 +1,44 @@
...
+    * ManagedNode
+        * value: ManagedNode
+        * Allowed Member types: Applications
+    * CliClient
+        * value: ManagedNode
+        * Allowed Member types: Applications
+* API Permissions
+    * User.Read (Microsoft Graph)
+* scope
+    * user_impersonation
```
Comment (ranweiler) [inline docs/AADEntitites.md:17]:
```suggestion
    * `user_impersonation`
```

File: docs/AADEntitites.md (hunk 21 above)
Comment (ranweiler) [reply docs/AADEntitites.md:15]: Found the docs for this:
```suggestion
    * _User.Read_ ([Microsoft Graph](https://docs.microsoft.com/en-us/graph/permissions-reference#user-permissions))
```

File: docs/AADEntitites.md [hunk 32]
```diff
@@ -0,0 +1,44 @@
...
+### OneFuzz CLI Service Principal
+Service principal linked to the OneFuzz CLI application registration.
+* name: `<instance_name>-cli`
+* Application Id: <OneFuzz CLI registration app_id>
+* User Assignment required: true
+* Permission
+    * CliClient (from OneFuzz Application registration)
+
+### Managed Node Service Principal
+This entity is available after the first deployment. This is the service principal associated with the user-assigned managed identity `<instance_name>-scalesetid`.
```
Comment (ranweiler) [inline docs/AADEntitites.md:39]: To match the usage below:
```suggestion
This entity is available after the first deployment. This is the service principal associated with the user-assigned managed identity `<instance_name>-<scaleset_id>`.
```

File: docs/AADEntitites.md [hunk 33]
```diff
@@ -0,0 +1,44 @@
...
+    * User.Read (Microsoft Graph)
+* scope
+    * user_impersonation
+* Authorized application:
+    * OneFuzz CLI registration
+
+### Onefuzz Application Service Principal
+Service principal linked to the OneFuzz application registration.
+* name: `<instance_name>`
+* Application Id: <OneFuzz Application registration app_id>
```
Comment (ranweiler) [inline docs/AADEntitites.md:24]: Without escaping or annotation, the angle-bracketed code gets treated as an (invalid) markup tag, and is absent when rendered:
```suggestion
* Application Id: _&lt;OneFuzz Application registration `app_id`&gt;_
```
There are a few similar spots below.

File: docs/AADEntitites.md [hunk 34]
```diff
@@ -0,0 +1,44 @@
...
+
+### OneFuzz CLI registration
+The registration for the command line interface.
+* name: `<instance_name>-cli`
+
+### OneFuzz CLI Service Principal
+Service principal linked to the OneFuzz CLI application registration.
+* name: `<instance_name>-cli`
+* Application Id: <OneFuzz CLI registration app_id>
```
Comment (ranweiler) [inline docs/AADEntitites.md:33]:
```suggestion
* Application Id: _&lt;OneFuzz CLI registration `app_id`&gt;_
```

File: docs/AADEntitites.md [hunk 35]
```diff
@@ -0,0 +1,44 @@
...
+* User Assignment required: true
```
Comment (ranweiler) [inline docs/AADEntitites.md:34]:
```suggestion
* User Assignment required: _true_
```

File: docs/AADEntitites.md [hunk 36]
```diff
@@ -0,0 +1,44 @@
...
+* Permission
+    * CliClient (from OneFuzz Application registration)
```
Comment (ranweiler) [inline docs/AADEntitites.md:36]:
```suggestion
    * _CliClient_ (from OneFuzz Application registration)
```

File: docs/AADEntitites.md [hunk 37]
```diff
@@ -0,0 +1,44 @@
...
+
+### Managed Node Service Principal
+This entity is available after the first deployment. This is the service principal associated with the user-assigned managed identity `<instance_name>-scalesetid`.
+
+* name: `<instance_name>-<scaleset_id>`
+* Service Principal
+    * Permission
+        * ManagedNode (from OneFuzz Application registration)
```
Comment (ranweiler) [inline docs/AADEntitites.md:44]:
```suggestion
        * _ManagedNode_ (from OneFuzz Application registration)
```

File: docs/AADEntitites.md [hunk 38]
```diff
@@ -0,0 +1,44 @@
+# Azure Active Directory Entities
+This document describes the configuration of entities create in Azure AD by our [deployment script](../src/deployment/deploy.py)
+
+### OneFuzz Application Registration
+This is the registration of the OneFuzz instance.
+* name : <instance_name>
+* app roles
+    * ManagedNode
```
Comment (ranweiler) [inline docs/AADEntitites.md:8]:
```suggestion
    * _ManagedNode_
```

File: docs/AADEntitites.md [hunk 39]
```diff
@@ -0,0 +1,44 @@
...
+        * value: ManagedNode
+        * Allowed Member types: Applications
+    * CliClient
```
Comment (ranweiler) [inline docs/AADEntitites.md:11]:
```suggestion
    * _CliClient_
```

**Review by ranweiler — APPROVED** [review]
Did one more pass, noticed an issue with rendered markdown and `<text like this>`.

File: docs/AADEntitites.md (hunk 33 above)
Comment (bmc-msft) [reply docs/AADEntitites.md:24]:
```suggestion
* Application Id: `<OneFuzz Application registration app_id>`
```

File: docs/AADEntitites.md (hunk 34 above)
Comment (bmc-msft) [reply docs/AADEntitites.md:33]:
```suggestion
* Application Id: `<OneFuzz CLI registration app_id>`
```

**Review by bmc-msft — APPROVED** [review]

**Pull request merged**